eng/common/scripts/modules/git-api-calls.psm1

```diff
@@ -0,0 +1,81 @@
+# Methods for interacting with the github or azure repo APIs
+# Retrieves the list of all tags that exist on the target repository
+# Changes 1: Pass full API URL instead of formulating it within the method
+# Changes 2: Extend method for git Repos
+function GetExistingTags($apiUrl, $repoType = "github", $headers = $null) {
+    try {
+      if ($repoType -eq "github")
+      {
+        return (Invoke-RestMethod -Method "GET" -Uri $apiUrl  ) | % { $_.ref.Replace("refs/tags/", "") }
+      }
+      else {
+        return (Invoke-RestMethod -Method "GET" -Uri $apiUrl -Headers $headers ) | % { $_.value.name.Replace("refs/tags/", "") }
+      }
+      
+    }
+    catch {
+      $statusCode = $_.Exception.Response.StatusCode.value__
+      $statusDescription = $_.Exception.Response.StatusDescription
+  
+      Write-Host "Failed to retrieve tags from repository."
+      Write-Host "StatusCode:" $statusCode
+      Write-Host "StatusDescription:" $statusDescription
+  
+      # Return an empty list if there are no tags in the repo
+      if ($statusCode -eq 404) {
+        return @()
+      }
+  
+      exit(1)
+    }
+}
+
+function FireAPIRequest($url, $method, $body = $null, $headers = $null, $repoType = "github") {
+    $attempts = 1
+  
+    while ($attempts -le 3) {
+      try {
+        return Invoke-RestMethod -Method $method -Uri $url -Body $body -Headers $headers
+      }
+      catch {
+        $response = $_.Exception.Response
+  
+        $statusCode = $response.StatusCode.value__
+        $statusDescription = $response.StatusDescription
+  
+        if ($statusCode) {
+          Write-Host "API request attempt number $attempts to $url failed with statuscode $statusCode"
+          Write-Host $statusDescription
+
+          if ($repoType -eq "github")
+          {
+            GitHubRateLimitInfo
+          }
+        }
+        else {
+          Write-Host "API request attempt number $attempts to $url failed with no statuscode present, exception follows:"
+          Write-Host $_.Exception.Response
+          Write-Host $_.Exception
+        }
+  
+        if ($attempts -ge 3) {
+          Write-Host "Abandoning Request $url after 3 attempts."
+          exit(1)
+        }
+  
+        Start-Sleep -s 10
+      }
+  
+      $attempts += 1
+    }
+}
+
+function GitHubRateLimitInfo() {
```

Comment:
You should pass in either response or headers instead of assuming they are globally set.